ivadomed/testing.py

```diff
@@ -243,11 +247,99 @@ def run_inference(test_loader, model, model_params, testing_params, ofolder, cud
                     gt_npy_list.append(np.array(gt_lst))
                     # Save merged labels with color
 
-                    if pred_undo.shape[0] > 1:
+                    if pred_undo.shape[0] > 1 and ofolder:
                         imed_utils.save_color_labels(pred_undo,
                                                      testing_params['binarize_prediction'] > 0,
                                                      batch['input_metadata'][smp_idx][0]['input_filenames'],
                                                      fname_pred.split(".nii.gz")[0] + '_color.nii.gz',
                                                      slice_axis)
 
     return preds_npy_list, gt_npy_list
+
+
+def threshold_analysis(model_path, ds_lst, model_params, testing_params, metric="dice", increment=0.1,
+                       fname_out="thr.png", cuda_available=True):
+    """Run a threshold analysis to find the optimal threshold on a sub-dataset.
+
+    Args:
+        model_path (str): Model path.
+        ds_lst (list): List of loaders.
+        model_params (dict): Model's parameters.
+        testing_params (dict): Testing parameters
+        metric (str): Choice between "dice" and "recall_specificity". If "recall_specificity", then a ROC analysis
+            is performed.
+        increment (float): Increment between tested thresholds.
+        fname_out (str): Plot output filename.
+        cuda_available (bool): If True, CUDA is available.
+
+    Returns:
+        float: optimal threshold.
+    """
+    if metric not in ["dice", "recall_specificity"]:
+        print('\nChoice of metric for threshold analysis: dice, recall_specificity.')
+        exit()
+
+    # Load model
+    model = torch.load(model_path)
+    # Eval mode
+    model.eval()
+
+    # List of thresholds
+    thr_list = list(np.arange(0.0, 1.0, increment))[1:]
+
+    # Init metric manager for each thr
+    metric_fns = [imed_metrics.recall_score,
+                  imed_metrics.dice_score,
+                  imed_metrics.specificity_score]
+    metric_dict = {thr: imed_metrics.MetricManager(metric_fns) for thr in thr_list}
+
+    # Load
+    loader = DataLoader(ConcatDataset(ds_lst), batch_size=8,
+                        shuffle=False, pin_memory=True, sampler=None,
+                        collate_fn=imed_loader_utils.imed_collate,
+                        num_workers=0)
+
+    # Run inference
+    preds_npy, gt_npy = run_inference(loader, model, model_params,
+                                      testing_params,
+                                      ofolder=None,
+                                      cuda_available=cuda_available)
+
+    # Make sure the GT is binarized
+    gt_npy = [threshold_predictions(gt, thr=0.5) for gt in gt_npy]
```

Comment:
I don't think a loop is necessary here

```suggestion
    gt_npy = threshold_predictions(gt_npy, thr=0.5)
```